design/legacy-block-storage.md

```diff
@@ -0,0 +1,91 @@
+# Legacy Block Storage with Rook
+
+
+## Overview
+
+### Value of legacy block storage support to Rook
+
+* In order to completely integrate with Kubernetes and utilize Rook's autonomous healing, scaling, and managing features, backend storage systems need to be software-based, distributed (clustered), and capable of being containerized. 
+Not all modern storage systems meet these requirements.
+* Many expensive and proprietary appliance storage silos are still providing services for Enterprise IT infrastructures.
+* Many companies only trust well-certified and time-proven storage solutions for their mission critical applications
+
+For transitioning into existing enterprise IT infrastructures, Rook will have a much easier time with support for legacy block storage systems.
+
+
+### Storage Resource Normalization
+
+The OpenStack Block Storage service (Cinder) provides a standardized architecture to aggregate the features and management of all kinds of block storage systems.
+It is the first service to have a standardized interface that is widely supported by most storage vendors including Dell EMC, IBM, NetApp, Hitachi, HPE, Pure Storage, VMware, etc.
+
+The major functions of the Cinder volume driver include:
+
+* Create/Delete Volume
+* Attach/Detach Volume
+* Extend Volume
+* Create/Delete Snapshot
+* Create Volume from Snapshot
+* Create Volume from Volume (clone)
+* Create Image from Volume
+* Volume Migration (host assisted)
+
+The following are some references for OpenStack Cinder:
+
+* [OpenStack Cinder](https://access.redhat.com/documentation/en/red-hat-enterprise-linux-openstack-platform/version-7/red-hat-enterprise-linux-openstack-platform-7-architecture-guide/chapter-1-components#comp-cinder)
+* [Cinder Driver Support Matrix](https://wiki.openstack.org/wiki/CinderSupportMatrix)
+
+In this project, we propose to leverage the open source project [OpenStack Cinder](https://github.com/openstack/cinder) as the unified storage connector that integrates legacy block storage with Rook.
+Through this implementation, we can quickly extend Rook's storage backend support to most modern block storage solutions that are already used in production today.
+
+
+## Architecture
+
+![legacy-block-storage-architecture-diagram](legacy-block-storage-architecture.png)
+
+### Operator
+
+* start and monitor the Cinder service pods
+* watch for state changes and apply the changes
+* monitor the health of external storage systems
+* manage custom resource definitions (CRDs) for pools
+
+
+### Agent
+
+* deployed on each Kubernetes nodes
+* handle attaching/detaching, mounting/unmounting volumes and formatting the file system
+
+
+## Details
+
+### lbs (legacy block storage) cluster CRD example
+
+```yaml
+apiVersion: lbs.rook.io/v1alpha1
+kind: Cluster
+metadata:
+  name: rook-lbs
+  namespace: rook-lbs
+spec:
+  cinderVersion:
+    image: openstack/cinder:v13.0.2-4
+  dataDirHostPath: /var/lib/rook
+  api:
```

Comment:
What is the difference between `api` and `scheduler`?

Reply:
api and scheduler are mapping to Cinder API and Cinder Scheduler services. Usually, API and scheduler services are deployed together. Maybe we should combine these two parameters into one.

Reply:
Agree, we should combine them into one.

> If I have 2 backends, does each backend get 3 API services and 3 scheduler services Or do the 2 backends share one or both of the services?

The 2 backends share the 3 API and 3 scheduler services.

> What are the effects of scaling either of the services? From my brief look at the Cinder docs, it seems that scaling the scheduler up would likely allow for more simultaneous clients, and the API translates the openstack api layer to the scheduler. I would imagine that scaling them together makes sense or scaling them with an n-to-one ratio perhaps if the scheduler has more throughput than the API (or vice versa?).

Scaling them together makes more sense.

> Is the API necessary? I suspect the answer is yes, but if the scheduler can be used directly that could reduce some complexity. It could also be possible but so much more complex to use that the API makes things less complex overall.

The API service is the official interface to communicate with Cinder and will be maintained with stability and compatibility between different Cinder versions. I would suggest we use the API instead of scheduler. Also, there may have some functions provided by API but not scheduler.